### DISCUSSION TITLE
Printed form of an exterior_algebra

### DISCUSSION THREAD

**Discussion by JohnAAbbott** [oscar-system/Oscar.jl#2238] [General]
I'd like to ask for comments/opinions about how an exterior algebra object should be printed out.
Of particular concern is that the printed form has size **quadratic** in the number of variables.
Here is an example with 8 variables:
```
julia> EA, e = exterior_algebra(QQ,8);

julia> EA
(PBW-algebra over Rational Field in e1, e2, e3, e4, e5, e6, e7, e8 with relations e2*e1 = -e1*e2, e3*e1 = -e1*e3, e4*e1 = -e1*e4, e5*e1 = -e1*e5, e6*e1 = -e1*e6, e7*e1 = -e1*e7, e8*e1 = -e1*e8, e3*e2 = -e2*e3, e4*e2 = -e2*e4, e5*e2 = -e2*e5, e6*e2 = -e2*e6, e7*e2 = -e2*e7, e8*e2 = -e2*e8, e4*e3 = -e3*e4, e5*e3 = -e3*e5, e6*e3 = -e3*e6, e7*e3 = -e3*e7, e8*e3 = -e3*e8, e5*e4 = -e4*e5, e6*e4 = -e4*e6, e7*e4 = -e4*e7, e8*e4 = -e4*e8, e6*e5 = -e5*e6, e7*e5 = -e5*e7, e8*e5 = -e5*e8, e7*e6 = -e6*e7, e8*e6 = -e6*e8, e8*e7 = -e7*e8)/two_sided_ideal(e1^2, e2^2, e3^2, e4^2, e5^2, e6^2, e7^2, e8^2)
```

It would be more compact and easier to read if it were printed out as something like
```
exterior_algebra over Rational Field in e1, e2, e3, e4, e5, e6, e7, e8
```

Maybe this is not so important: how often does one want to print out an exterior algebra object?  how often does one use an exterior algebra with a large number of variables?

**Comment by lkastner** [comment]
There just was some addition to the [styleguide on printing](https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/#Printing-in-Oscar). It looks to me as if your second variant would be the "one line" form and the first one would be "detailed". Maybe "supercompact" would be something like `exterior_algebra over Rational Field`?

cc @simonbrandhorst

**Reply by JohnAAbbott** [reply]
@lkastner  Thanks for the helpful link!

**Reply by JohnAAbbott** [reply]
Now I am unsure how to reconcile the styleguide on printing with the recommendation at https://docs.oscar-system.org/stable/AbstractAlgebra/ring_interface/  in the subsection about _expressify_.   My current opinion is that the styleguide seems nicer (but also requires more effort to implement).

**Reply by lkastner** [reply]
In general I'd think that things in Oscar.jl should adhere to the styleguide. However I would be even happier if there were no contradictions, so maybe we should talk to AbstractAlgebra people to resolve.

**Reply by JohnAAbbott** [reply]
Yes, I think we need to talk to more people --- who would be the best AbstractAlgebra person?
The recommendation (at the link i posted above) to use _expressify_ claimed advantages when producing non-plain-text output (_e.g._ LaTeX or HTML); no doubt the idea is that the expression is an intermediate, structured form which can easily be serialized for different formats (and should automatically ensure consistency _e.g._ of indentation).  But who does the serialization of the result of `expressify`?  Is it just printed by Julia as a Julia value (with sometimes unfortunate results)?
The style guide mentions 3 levels of detail in the output: `detailed`, `one line` (why not `one_line`?), and `supercompact`.  Presumably this is relevant only for plain-text output -- it seems irrelevant for LaTeX or HMTL.

As a point of reference, I created two `weyl_algebra` values then printed the list/vector containing them:
```
julia> [D, Dab]
2-element Vector{PBWAlgRing{QQFieldElem, Singular.n_Q}}:
 PBW-algebra over Rational Field in x, y, dx, dy with relations y*x = x*y, dx*x = x*dx + 1, dy*x = x*dy, dx*y = y*dx, dy*y = y*dy + 1, dy*dx = dx*dy
 PBW-algebra over Rational Field in a, b, da, db with relations b*a = a*b, da*a = a*da + 1, db*a = a*db, da*b = b*da, db*b = b*db + 1, db*da = da*db

```
In my estimation the readability of this output is poor.  Was the `one line` detail-level active?

**Reply by simonbrandhorst** [reply]
Indeed the readability is poor. The new printing guidelines are pretty new and so far there is almost no type adhering to them.
But there is an abstract algebra pull request making a start.

**Comment by simonbrandhorst** [comment]
Indeed, we should print it as
```
julia> EA
Exterior algebra in 8 variables e1, e2, e3, e4, e5, ... , e8
  over Rational field
```

```
julia> [EA, EA]
2-element Vector{PBWAlgQuo{QQFieldElem, Singular.n_Q}}
 Exterior algebra in 8 variables over Rational field
 Exterior algebra in 8 variables over Rational field
```
with `supercompact`  as `Exterior algebra`. 

For general PBW-algebras, we could go with something along the lines of 
(not sure if we want to display the relations.)
```
julia> EA
Quotient of 
  PBW-algebra in 8 variables over Rational field 
    with relations (e2*e1 = -e1*e2, e3*e1 = -e1*e3, e4*e1 = -e1*e4, ...., e8*e7 = -e7*e8)
  by two_sided_ideal(e1^2, e2^2, e3^2, e4^2, e5^2, e6^2, e7^2, e8^2)
```

```
julia> [EA,EA]
2-element Vector{PBWAlgQuo{QQFieldElem, Singular.n_Q}}
 Quotient of PBW-algebra in 8 variables over Rational field
 Quotient of PBW-algebra in 8 variables over Rational field
```

**Reply by JohnAAbbott** [reply]
@simonbrandhorst  Your suggestion for (quotient of) PBW algebras offers better readability than the current implementation.
This should be not too hard to implement.

I'm still wondering how to achieve your proposal for exterior algebras:  currently they are represented as quotients of PBW algebras (and the history of how they were created is not retained).  A similar suggestion no doubt applies to Weyl algebras.

Philosophical question: _should the way an object is printed be influenced by how it was created?  Or more generally by its history?_

**Reply by JohnAAbbott** [reply]
@simonbrandhorst  I have some "niggling detail" questions.
You proposed the `detailed` output should look like this:
```
julia> EA
Exterior algebra in 8 variables e1, e2, e3, e4, e5, ... , e8
  over Rational field
```
Why is `Rational field` on a separate line but the variables are not?
Why do you print the variables first then print the coefficient ring?
Did you actually mean that the output should contain the ellipsis `...` (at least when there are many variables)?

Here is an alternative possible output:
```
julia> EA
Exterior algebra
  over Rational field
  in 8 variables e1, e2, e3, e4, e5, ...
```
I'm not claiming my version is better; I just wanted to give an alternative example.  The coefficient ring is printed before the variables, and just the first few variables are printed (with a final ellipsis to indicate that there are more, but without giving the name of the last variable).   The `one_line` version could then just be the first part:
```
Exterior algebra over Rational field in 8 variables
```

**Reply by simonbrandhorst** [reply]
Not giving the last variable after the ellipsis looks better indeed. 

As for printing the variables first, this is to make sure that they stay close to `Exterior Algebra` which is where they live. 
If they are farther away, I think it is a bit harder to read. 
```
Exterior algebra
  over Polynomial ring in 5 variables over QQ
  in 8 variables e1, e2, e3, e4, e5, ...
```
versus
```
Exterior algebra in 8 variables e1, e2, e3, e4, e5, ...
  over Polynomial ring in 5 variables over QQ
```
Your suggestion 
```
Exterior algebra 
  in 8 variables e1, e2, e3, e4, e5, ...
  over Polynomial ring in 5 variables
```
looks okay too. We should just be consistent.

**Reply by JohnAAbbott** [reply]
Good comment about the placement of `in NN variables` -- this is especially relevant for the `one line` version.
In the documentation [https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/#Printing-in-Oscar](url)
the notes about `one line` point out that commas should be inside brackets.  While this is not explicitly required/recommended for the `detailed` version, I presume for visual consistency one should use brackets there too.  This would lead us to:
```
Exterior algebra 
  in 8 variables (e1, e2, e3, e4, e5, ...)
  over Polynomial ring in 5 variables
```
Personally I think it looks OK with the variable names inside brackets.

**Reply by JohnAAbbott** [reply]
@simonbrandhorst  Earlier you asked whether we want to print out the relations.  In general, we have no choice (since they are an essential defining characteristic).  However there are n^2 of them.  I suppose we just have to accept that a fully detailed printed form is going to be large (and practically unreadable); using ellipsis to keep it shorter would defeat the purpose of a "fully detailed" description.
Not sure if it is a good analogy: Groebner bases are often large (and so "unreadable"); so one quickly learns not to print them out.

**Reply by simonbrandhorst** [reply]
@thofma  is finalizing our printing in https://github.com/Nemocas/AbstractAlgebra.jl/pull/1320
Best wait until this is through and then make the PBW-algebras consistent with that. 
Some of our choices may not be perfect and depend on personal taste, but let's stick to them anyways for some months.  
If after some time we have a strong opinion that a parent should print differently, we can still improve the printing.  For now let's just be consistent.

**Comment by wdecker** [comment]
@JohnAAbbott There are two suggestions here which are both fine:
The first one applies to the exterior algebra (no history needed). Similarly for the Weyl algebra. Please inplement.
The second one applies to all quotients of PBW algebras except the exterior algebra. Similarly for the PBW algebras other than the Weyl algebra. Please implement. 

@simonbrandhorst With respect to the relations, but the questions is more general: Can we print dynamically in the sense that if the output exceeds a certain size, then only something very compact is written (like nn relations instead of relations (...)

**Reply by simonbrandhorst** [reply]
> @simonbrandhorst With respect to the relations, but the questions is more general: Can we print dynamically in the sense that if the output exceeds a certain size, then only something very compact is written (like nn relations instead of relations (...)

Sounds reasonable to me and on the technical side it is doable.

**Reply by JohnAAbbott** [reply]
@wdecker  To achieve special printing for exterior algebras and Weyl algebras, I believe it will be practically necessary for the internal representation to store the information that it was created as a WA or EA: it would probably be not too costly to check whether a PBW algebra is actually a Weyl algebra (just check n^2 multplication relations), but checking that a quotient is an exterior algebra could be costly if it is not one (since we need to perform an equality test of ideals, in addition to checking the multiplication relations).
The use of `expressify` for printing seems also to be an obstacle (or aggravating consideration).

I am assuming that we hope to achieve something like the following:
```
julia> D, (x, y, dx, dy) = weyl_algebra(QQ, ["x", "y"]);
julia> D
Weyl algebra
  over Rational field
  with 2 variables x, y
```
Notice that internally there are 4 variables, but printing just refers to the non-derivation ones

**Reply by simonbrandhorst** [reply]
@JohnAAbbott 
Yes, this can be stored. For instance we could use the `@attr` macro and define an internal function `is_created_as_weyl_algebra`.

**Reply by JohnAAbbott** [reply]
Yesterday I pushed a revision to the exterior algebra code; it also revised the printing of Weyl algebras.
Printing is still via _expressify_, and **does not respect** `detailed`, `one_line` and `supercompact`.

**Comment by fieker** [comment]
On Fri, Apr 14, 2023 at 04:53:33AM -0700, JohnAAbbott wrote:
> @simonbrandhorst  I have some "niggling detail" questions.
> You proposed the `detailed` output should look like this:
> ```
> julia> EA
> Exterior algebra in 8 variables e1, e2, e3, e4, e5, ... , e8
>   over Rational field
> ```
> Why is `Rational field` on a separate line but the variables are not?
Because of recursion, we tried it othewise:
 ext algebra
   over mpoly ring in 4 vars ...
     over rational field in 8 vars ...
here it is not clear where the variables live
